File: 105_ConstructBinaryTreefromPreorderandInorderTraversal.java
class Solution {
    HashMap<Integer, Integer> map = new HashMap<>();
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        for (int i = 0; i < inorder.length; i++) {
            map.put(inorder[i], i);
        }
        return dfs(preorder, inorder, 0, inorder.length - 1, 0);
    }

    private TreeNode dfs(int[] preorder, int[] inorder, int l, int r, int rootID) {
        if (l > r) {
            return null;
        }
        int val = preorder[rootID];
        TreeNode node = new TreeNode(val);
        int p = map.get(val);
        node.left = dfs(preorder, inorder, l, p - 1, rootID + 1);
        node.right = dfs(preorder, inorder, p + 1, r, rootID + 1 + (p - l));
        return node;
    }
}

File: c++/105_ConstructBinaryTreefromPreorderandInorderTraversal.cpp
class Solution {
public:
    unordered_map<int, int> mp;
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        for (int i = 0; i < inorder.size(); i++) mp[inorder[i]] = i;
        return build(preorder, inorder, 0, preorder.size() - 1, 0, inorder.size() -1);
    }
    
    TreeNode* build(vector<int>& preorder, vector<int>& inorder, int pl, int pr, int il, int lr) {
        if (pl > pr) return nullptr;
        auto root = new TreeNode(preorder[pl]);
        int k = mp[preorder[pl]];
        root->left = build(preorder, inorder, pl+1, pl+k-il, il, k-1);
        root->right = build(preorder, inorder, pl+k-il+1, pr, k+1, pr);
        return root;
    }
};
